### PR TITLE
fix: propagate null-valued records in repartition

### PR DESCRIPTION
**Pull request by vcrfxia**
### Description 

Fixes https://github.com/confluentinc/ksql/issues/6646

The current mapper used in repartitions builds the new key by evaluating the generated expression on the value. If the value is null, the mapper returns an empty key rather than the original key. More intuitive behavior is to detect when the partition by expression depends only on key column(s) and evaluate the new key on the key columns in these cases, allowing null values to be propagated. This is needed to enable the propagation of tombstones in https://github.com/confluentinc/ksql/pull/6635

### Testing done 

Unit + QTT.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/StructKeyUtil.java [hunk 1]
```diff
@@ -58,7 +58,7 @@ public static KeyBuilder keyBuilder(final ColumnName name, final SqlType type) {
   }
 
   @SuppressWarnings("unchecked")
-  public static List<?> asList(final Object key) {
+  public static List<Object> asList(final Object key) {
```
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/StructKeyUtil.java:61]: Is there a reason this was previously `List<?>` rather than `List<Object>`? I don't fully understand the meaning of this change, but it lets us avoid an unchecked cast downstream.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ColumnExtractor.java [hunk 2]
```diff
@@ -13,7 +13,7 @@
  * specific language governing permissions and limitations under the License.
  */
 
-package io.confluent.ksql.analyzer;
+package io.confluent.ksql.execution.util;
```
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ColumnExtractor.java:16]: I don't love the package this was moved to but couldn't think of anywhere better. Suggestions welcome!

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java [hunk 3]
```diff
@@ -202,6 +217,23 @@ private static LogicalSchema buildSchema(
...
+    return new PartitionByExpressionEvaluator(
+        row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg),
+        isKeyExpression
+    );
+  }
+
+  private static class PartitionByExpressionEvaluator {
+
+    private final Function<GenericRow, Object> evaluator;
+    private final boolean acceptsKey;
```
Comment (agavra) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java:229]:
```suggestion
    private final boolean operatesOnKeyOnly;
```
A little hard for me to understand it inline without context on what this PR does:
```java
      final Object newKey = evaluator.evaluator.apply(
          evaluator.acceptsKey
          ? GenericRow.fromList(StructKeyUtil.asList(k))
          : v
      );
```

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java [hunk 4]
```diff
@@ -202,6 +217,23 @@ private static LogicalSchema buildSchema(
     final String errorMsg = "Error computing new key from expression "
         + expressionMetadata.getExpression();
 
-    return row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg);
+    return new PartitionByExpressionEvaluator(
+        row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg),
```
Comment (agavra) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java:221]: I suspect this isn't the only place where this bug happens (a null value doesn't get properly passed into an `ExpressionEvaluator` 🤔). I had a PR at one point going to try to have `ExpressionMetadata#evaluate` take in both a key and a value instead of just a `GenericRow` which would have prevented this issue. Unfortunately, I couldn't get joins to work because the joiner doesn't actually have access to the key 😭 

tldr, I think it would be good if we encapsulate as much of this "put the keys and only the keys into the generic row if the row was null" into the `PartitionByExpressionEvaluator` and have that implement `BiFunction<Struct, GenericRow, Object>` so at least the part that creates the key is all in the same place that calls the evaluator:
```java
      final Object newKey = evaluator.evaluator.apply(
          evaluator.acceptsKey
          ? GenericRow.fromList(StructKeyUtil.asList(k))
          : v
      );
```

I would also just pass in `expressionMetadata` into the `PartitionByExpressionEvaluator` instead of passing it in as a lambda so that it's centralized and we don't have to track down where it's being called and what's getting passed into it

**Review by agavra — APPROVED** [review]
Superb sleuthing 🕵️ some minor comments inline but I think we can sneak this by without worrying about backwards compatibility too much (famous last words)

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java [hunk 5]
```diff
@@ -87,12 +90,19 @@ public static PartitionByParams build(
       // is already present in the current value
       mapper = (k, v) -> new KeyValue<>(null, v);
     } else {
-      final Function<GenericRow, Object> evaluator = buildExpressionEvaluator(
+      final Set<? extends ColumnReferenceExp> partitionByCols =
+          ColumnExtractor.extractColumns(partitionBy);
+      final boolean isKeyExpression = partitionByCols.stream()
+          .map(ColumnReferenceExp::getColumnName)
+          .allMatch(sourceSchema::isKeyColumn);
```
[the remark this answers is not in the document]
Comment (vcrfxia) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java:97]: Not sure I understand. The validation that all key columns are selected happens elsewhere. This piece of code just determines whether only the old key is needed to compute the new key (partition by expression) or if the value is needed as well. If the former, we can propagate null-valued records, else we can't.

[the remark this answers is not in the document]
Comment (vcrfxia) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java:97]: Renamed `isKeyColumn` -> `partitionByInvolvesKeyColsOnly` to clarify.

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java (hunk 4 above)
Comment (vcrfxia) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java]: Ah, thanks for the suggestion -- much cleaner!

What's the benefit of implementing `BiFunction<Object, GenericRow, Object>` rather than simply have a method? I've currently got
```
  Object evaluate(final Object key, final GenericRow value) {
      final GenericRow row = evaluateOnKeyOnly
          ? GenericRow.fromList(StructKeyUtil.asList(key))
          : value;
      return expressionMetadata.evaluate(row, null, logger, errorMsg);
    }
```

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java (hunk 3 above)
Comment (vcrfxia) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java]: I think this is addressed by your suggestion below? I went with `evaluateOnKeyOnly` though I don't feel strongly.

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java (hunk 4 above)
Comment (agavra) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java]: there is none, I think that's great (especially if nowhere else expects the BiFunction)

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java:97]: ah! that makes more sense. I had misread what this code was doing.
